lib/SILOptimizer/LoopTransforms/LICM.cpp

```diff
@@ -476,23 +480,26 @@ static void analyzeBeginAccess(BeginAccessInst *&BI,
     return;
   }
 
-  // find matching end access:
+  // find matching end accesses:
+  HoistSet matchingEnds;
   auto matchingEndPred = [&](EndAccessInst *EI) {
     return EI->getBeginAccess() == BI;
   };
-  auto matchingEnd =
-      std::find_if(EndAccesses.begin(), EndAccesses.end(), matchingEndPred);
-  if (matchingEnd == EndAccesses.end()) {
-    // no matching end within the loop
-    return;
+  for (auto matchingEnd = std::find_if(EndAccesses.begin(), EndAccesses.end(),
```

Comment:
With this approach, I'm afraid you could miss some of the end_access instructions because they weren't visited during the loop analysis. It would be better to scan the begin's use list gathering all it's end_access instructions, and verifying that they are all inside the loop.

Reply:
I have made all the changes in this review except this one: I will either bail or add support + test per our discussion offline. looking into it...